### PR TITLE
Introduce component serializers

### PR DESCRIPTION
**Pull request by kashike**

### CODE REVIEW

File: src/main/java/net/kyori/text/serializer/ComponentSerializers.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * A collection of component serializers.
+ */
+public final class ComponentSerializers {
+
+  public static final ComponentSerializer<Component, String> GSON = GsonComponentSerializer.INSTANCE;
+  /**
+   * @deprecated legacy
+   */
+  @Deprecated
```
Comment (jamierocks) [inline src/main/java/net/kyori/text/serializer/ComponentSerializers.java]: no need for this to be deprecated

sure the spec underlying the code is deprecated, but this is just going to make peoples ides moan at them, although perhaps it should if they use legac.....

carry on

File: src/main/java/net/kyori/text/serializer/LegacyComponentSerializer.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import javax.annotation.Nonnull;
+
+/**
+ * A legacy component serializer.
+ *
+ * @param <C> the component type
+ * @deprecated legacy
+ */
+@Deprecated
+public interface LegacyComponentSerializer<C extends Component> extends ComponentSerializer<C, String> {
```
Comment (jamierocks) [inline src/main/java/net/kyori/text/serializer/LegacyComponentSerializer.java]: probably just dperecating the class will do

File: src/main/java/net/kyori/text/serializer/ComponentSerializers.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+package net.kyori.text.serializer;
+
+import net.kyori.text.Component;
+
+/**
+ * A collection of component serializers.
+ */
+public final class ComponentSerializers {
+
+  public static final ComponentSerializer<Component, String> GSON = GsonComponentSerializer.INSTANCE;
```
Comment (kashike) [inline src/main/java/net/kyori/text/serializer/ComponentSerializers.java]: `GSON` or `JSON`?

Comment (connorhartley) [inline src/main/java/net/kyori/text/serializer/ComponentSerializers.java]: Difficult question... I would say... keep it how it is.

Comment (lucko) [inline src/main/java/net/kyori/text/serializer/ComponentSerializers.java]: I would take the approach Sponge has. 

LEGACY_CHAR and JSON. 

Users of text shouldn't need to worry about how the method is implemented.

Comment (jamierocks) [inline src/main/java/net/kyori/text/serializer/ComponentSerializers.java]: JSON

Comment (mikroskeem) [inline src/main/java/net/kyori/text/serializer/ComponentSerializers.java]: Yes JSON would make more sense

File: src/main/java/net/kyori/text/serializer/ComponentSerializers.java (hunk 1 above)
Comment (lucko) [inline src/main/java/net/kyori/text/serializer/ComponentSerializers.java]: Idk, isn't just being called "LEGACY" enough? 

Deprecation warnings are annoying :(

File: src/main/java/net/kyori/text/serializer/GsonComponentSerializer.java [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+import net.kyori.text.format.TextDecoration;
+
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public enum GsonComponentSerializer implements ComponentSerializer<Component, String>, JsonDeserializer<Component>, JsonSerializer<Component> {
```
Comment (lucko) [inline src/main/java/net/kyori/text/serializer/GsonComponentSerializer.java]: Any reason why this is public?

Shouldn't the instance be obtained from the ComponentSerializers class?

Comment (kashike) [inline src/main/java/net/kyori/text/serializer/GsonComponentSerializer.java]: To allow registration on custom `Gson` instances.

**Pull request merged**